File: src/discord/directoryBrowser.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as os from 'node:os';
import type { ComponentRow, EmbedSpec, SelectSpec } from './ports.js';
import { t } from './i18n.js';

// Filesystem folder navigation used by the channel wizard (§9): list dirs, go up,
// go into, select the current path. By default (no allowedRoots) the admin driver can
// browse ANYWHERE up to the filesystem root '/', so the session cwd can be a project on
// any volume (e.g. /Volumes/<other-drive> on macOS): ⬆ up is enabled at every path
// except '/' itself. When allowedRoots IS supplied, navigation stays CONFINED within
// them (the original A5 path-escape guard, still used by callers that want a bounded
// picker). NOTE: this governs only CHOOSING the cwd; the per-session file-access
// confinement (attach/download realpath-confined to the chosen cwd) is a SEPARATE
// mechanism and is unaffected. Pure FS + string logic — no discord.js — so it is
// unit-testable against temp dirs; the render() output is a plain component spec that
// 7b maps onto discord.js.
//
// custom_id scheme (parsed by 7b's interactionRouter):
//   dir:into    (string-select; value = child folder name)
//   dir:up      (button — go to the parent folder)
//   dir:here    (button — select the current folder / Session Start)
//   dir:resume  (button — start the resume-session flow for the current folder)
//   dir:create  (button — open the create-folder modal in the current folder)
//   dir:manual  (button — open the manual absolute-path modal; jumps the browser there)
//   dir:panel   (button — open a native folder picker on the host; jumps the browser there)
//   cancel      (button — cancel the wizard)

// Discord select limits (A4D MAX_SELECT_OPTIONS / label length).
const MAX_SELECT_OPTIONS = 25;
const MAX_LABEL_LENGTH = 95;

export interface DirectoryBrowserOptions {
  // Absolute directories the user may browse within. Any navigation target must be
  // inside at least one root. When OMITTED/empty the browser is UNBOUNDED: the driver
  // can navigate up to the filesystem root '/' and into any volume (the admin picks the
  // session cwd anywhere). Supply roots only for a deliberately bounded picker.
  allowedRoots?: string[];
  // Where to start; must resolve inside an allowed root (bounded mode) or it is clamped
  // to the first root. Defaults to the first allowed root, or to home when unbounded.
  startPath?: string;
  // Offer the 🖥️ native host-side folder picker button (dir:panel). The router sets
  // this only when a picker implementation is wired (macOS host, §folderPanel.ts) —
  // the browser stays pure and platform-agnostic; it just renders one more button.
  nativePanel?: boolean;
}

export class DirectoryBrowser {
  // The confinement roots, or null when the browser is unbounded (browse anywhere up to
  // '/'). Kept separate from `current` so navigation checks pick the right rule.
  private readonly roots: string[] | null;
  private readonly nativePanel: boolean;
  private current: string;

  constructor(options: DirectoryBrowserOptions = {}) {
    const bounded = Boolean(options.allowedRoots && options.allowedRoots.length > 0);
    this.roots = bounded ? options.allowedRoots!.map((r) => path.resolve(r)) : null;
    this.nativePanel = options.nativePanel ?? false;
    const fallbackStart = this.roots ? this.roots[0] : os.homedir();
    const start = options.startPath ? path.resolve(options.startPath) : fallbackStart;
    this.current = this.confine(start) ? start : fallbackStart;
  }

  // The folder currently in view.
  cwd(): string {
    return this.current;
  }

  // Names of immediate subdirectories of the current folder (sorted, capped). Unreadable
  // dirs and non-directory entries are skipped; a permission error yields an empty list.
  // Hidden (dot) folders ARE listed, but sorted AFTER normal folders so they can't push
  // real project folders past the MAX_SELECT_OPTIONS cap; each group stays alphabetical.
  listChildren(): string[] {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(this.current, { withFileTypes: true });
    } catch {
      return [];
    }
    return entries
      .filter((e) => e.isDirectory())
      .map((e) => e.name)
      .sort((a, b) => {
        const ah = a.startsWith('.'), bh = b.startsWith('.');
        if (ah !== bh) return ah ? 1 : -1; // non-dot first, dot last
        return a.localeCompare(b);
      })
      .slice(0, MAX_SELECT_OPTIONS);
  }

  // Descend into a child folder. No-ops (returns false) if the child does not exist,
  // is not a directory, or would escape an allowed root.
  into(childName: string): boolean {
    const target = path.resolve(this.current, childName);
    if (!this.confine(target)) return false;
    if (!isDirectory(target)) return false;
    this.current = target;
    return true;
  }

  // Go up one level. No-ops (returns false) at the filesystem root ('/', where
  // dirname(p) === p) and — in bounded mode — at an allowed-root boundary. In unbounded
  // mode the only stop is '/', so the driver can reach any volume above the start dir.
  up(): boolean {
    const parent = path.dirname(this.current);
    if (parent === this.current) return false;
    if (!this.confine(parent)) return false;
    this.current = parent;
    return true;
  }

  // Jump directly to an absolute path typed via the 📝 manual-path option. No-ops
  // (returns false) if the path does not exist, is not a directory, or would escape an
  // allowed root — the SAME confinement rule as into()/up(). On success the current view
  // moves there so ✅ Start selects it (the driver still confirms; no auto-start).
  goTo(target: string): boolean {
    const resolved = path.resolve(target);
    if (!this.confine(resolved)) return false;
    if (!isDirectory(resolved)) return false;
    this.current = resolved;
    return true;
  }

  // Select the current folder as the session cwd. Returns its absolute path.
  select(): string {
    return this.current;
  }

  // True when navigation to `target` is permitted. Unbounded mode (roots === null)
  // allows any absolute path (the whole filesystem, up to '/'). Bounded mode requires
  // `target` to be the same as, or nested under, at least one allowed root; path.relative
  // is used so /ws is not fooled by /ws-evil.
  private confine(target: string): boolean {
    const resolved = path.resolve(target);
    if (this.roots === null) return true;
    return this.roots.some((root) => isWithin(root, resolved));
  }

  // Build the interactive spec (embed + component rows) for the current view, A4D-style:
  // the embed shows the CURRENT path + how-to guidance, a select menu lists subfolders,
  // and a single action row mirrors A4D's folder-step buttons —
  // ⬆ 상위(Parent) · ✅ 시작(Session Start) · 세션 재개(Resume) · 📁 폴더 만들기(Create) ·
  // 취소(Cancel) — driving navigation, selection, resume, create, and cancel. Two rows
  // total (select + buttons), well within Discord's 5-row-per-message limit. Pure data;
  // 7b turns it into a discord.js reply/update.
  render(): { embed: EmbedSpec; rows: ComponentRow[] } {
    const children = this.listChildren();
    const embed: EmbedSpec = {
      title: t('wizard.step.folder'),
      description: t('dir.guide') + '\n\n' + t('dir.current') + ': `' + this.current + '`',
    };
    const select: SelectSpec = {
      type: 'select',
      customId: 'dir:into',
      placeholder: children.length > 0 ? t('dir.select') : t('dir.empty'),
      options:
        children.length > 0
          ? children.map((name) => ({ label: clip(name, MAX_LABEL_LENGTH), value: name }))
          : [{ label: t('dir.empty'), value: '__none__' }],
    };
    const rows: ComponentRow[] = [
      { components: [select] },
      {
        components: [
          { type: 'button', customId: 'dir:up', label: t('dir.up'), style: 'secondary', disabled: !this.canGoUp() },
          { type: 'button', customId: 'dir:here', label: t('dir.here'), style: 'success' },
          { type: 'button', customId: 'dir:resume', label: t('dir.resume'), style: 'primary' },
          { type: 'button', customId: 'dir:create', label: t('dir.create'), style: 'secondary' },
          { type: 'button', customId: 'cancel', label: t('wizard.cancel'), style: 'secondary' },
        ],
      },
      {
        // A separate row (the button row above is already at Discord's 5-button limit):
        // type an absolute path instead of clicking down to it (also handy on mobile /
        // for deep paths). Opens a modal; on submit the browser jumps there. When a
        // native host-side picker is wired (macOS), dir:panel joins it — pick the
        // folder in a real Finder open-panel instead of typing.
        components: [
          { type: 'button', customId: 'dir:manual', label: t('dir.manual'), style: 'secondary' },
          ...(this.nativePanel
            ? [{ type: 'button' as const, customId: 'dir:panel', label: t('dir.panel'), style: 'secondary' as const }]
            : []),
        ],
      },
    ];
    return { embed, rows };
  }

  private canGoUp(): boolean {
    const parent = path.dirname(this.current);
    return parent !== this.current && this.confine(parent);
  }
}

function isDirectory(p: string): boolean {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function isWithin(root: string, child: string): boolean {
  const rel = path.relative(root, child);
  return rel === '' || (!rel.startsWith('..') && !path.isAbsolute(rel));
}

function clip(s: string, max: number): string {
  return s.length <= max ? s : s.slice(0, max - 1) + '…';
}

File: swift/Sources/DiscordAgentBridge/Session/DirectoryBrowser.swift
import Foundation

// MARK: - W11-b2: pure filesystem folder browser (TS directoryBrowser.ts parity)
//
// List dirs, go up/into, select cwd. No Discord types. Optional allowedRoots confine;
// when omitted/empty the browser is unbounded (navigate to filesystem root '/').
// Session file-access confinement is separate (Confinement.swift) and unaffected.
//
// custom_id scheme (DabMain / wizard handle):
//   dir:into   string-select value = child folder name
//   dir:up     button — parent
//   dir:here   button — commit cwd / advance wizard
//   dir:create button — open create-folder modal (handled in DabMain)
//   dir:manual button — open absolute-path modal (handled in DabMain)
//   dir:panel  button — native host picker when `nativePanel` (handled in DabMain)
//   dir:resume button — start Resume Session flow (ResumeWizard; handled in DabMain)
//   cancel     button — cancel wizard

private let maxSelectOptions = 25
private let maxLabelLength = 95

public struct DirectoryBrowserOptions: Sendable, Equatable {
    /// Absolute directories the user may browse within. Empty/nil = unbounded.
    public var allowedRoots: [String]
    /// Start path; clamped to first root when out of bounds (bounded mode).
    public var startPath: String?
    /// Offer native host picker button (slice3). Default off.
    public var nativePanel: Bool

    public init(
        allowedRoots: [String] = [],
        startPath: String? = nil,
        nativePanel: Bool = false
    ) {
        self.allowedRoots = allowedRoots
        self.startPath = startPath
        self.nativePanel = nativePanel
    }
}

public final class DirectoryBrowser: @unchecked Sendable {
    /// Confinement roots, or nil when unbounded.
    private let roots: [String]?
    private let nativePanel: Bool
    private var current: String

    public init(options: DirectoryBrowserOptions = DirectoryBrowserOptions()) {
        let bounded = !options.allowedRoots.isEmpty
        self.roots = bounded
            ? options.allowedRoots.map { Self.resolvePath($0) }
            : nil
        self.nativePanel = options.nativePanel
        let fallbackStart = self.roots?.first ?? NSHomeDirectory()
        let start: String
        if let sp = options.startPath {
            start = Self.resolvePath(sp)
        } else {
            start = fallbackStart
        }
        self.current = Self.confine(start, roots: self.roots) ? start : fallbackStart
    }

    /// Convenience: bounded/unbounded from roots + start.
    public convenience init(
        allowedRoots: [String] = [],
        startPath: String? = nil,
        nativePanel: Bool = false
    ) {
        self.init(options: DirectoryBrowserOptions(
            allowedRoots: allowedRoots,
            startPath: startPath,
            nativePanel: nativePanel
        ))
    }

    public func cwd() -> String { current }

    /// Immediate subdirectories: non-dot first (alpha), then dot folders (alpha); cap 25.
    public func listChildren() -> [String] {
        let fm = FileManager.default
        guard let names = try? fm.contentsOfDirectory(atPath: current) else { return [] }
        var dirs: [String] = []
        for name in names {
            let full = (current as NSString).appendingPathComponent(name)
            var isDir: ObjCBool = false
            guard fm.fileExists(atPath: full, isDirectory: &isDir), isDir.boolValue else { continue }
            dirs.append(name)
        }
        dirs.sort { a, b in
            let ah = a.hasPrefix("."), bh = b.hasPrefix(".")
            if ah != bh { return !ah } // non-dot first
            return a.localizedStandardCompare(b) == .orderedAscending
        }
        if dirs.count > maxSelectOptions {
            return Array(dirs.prefix(maxSelectOptions))
        }
        return dirs
    }

    /// Descend into a child. false if missing / not a dir / escapes roots.
    @discardableResult
    public func into(_ childName: String) -> Bool {
        let target = Self.resolvePath(current, childName)
        guard Self.confine(target, roots: roots) else { return false }
        guard Self.isDirectory(target) else { return false }
        current = target
        return true
    }

    /// Go up one level. false at filesystem root or root boundary (bounded).
    @discardableResult
    public func up() -> Bool {
        let parent = (current as NSString).deletingLastPathComponent
        if parent == current { return false }
        guard Self.confine(parent, roots: roots) else { return false }
        current = parent
        return true
    }

    /// Jump to absolute path (manual path / native panel / tests). Same confinement as into/up.
    @discardableResult
    public func goTo(_ target: String) -> Bool {
        let resolved = Self.resolvePath(target)
        guard Self.confine(resolved, roots: roots) else { return false }
        guard Self.isDirectory(resolved) else { return false }
        current = resolved
        return true
    }

    /// Create a direct child folder under cwd (modal submit). Safe name + confine; optional enter.
    public func createChild(_ name: String, enter: Bool = true) -> DirCreateResult {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard isSafeFolderName(trimmed) else { return .invalidName }
        let parent = current
        let target = (parent as NSString).appendingPathComponent(trimmed)
        // Defense in depth: parent of target must still be current (no separator tricks).
        guard (target as NSString).deletingLastPathComponent == parent else { return .invalidName }
        guard Self.confine(target, roots: roots) else { return .escaped }
        do {
            try FileManager.default.createDirectory(atPath: target, withIntermediateDirectories: true)
        } catch {
            return .failed(String(describing: error))
        }
        if enter {
            _ = into(trimmed)
        }
        return .ok(path: target)
    }

    /// Select current folder as session cwd.
    public func select() -> String { current }

    /// Folder-step UI (select + up/here/create/cancel + manual/[panel]). Pure data → DiscordBM.
    public func render() -> WizardView {
        let children = listChildren()
        let title = I18n.t("wizard.step.folder")
        // TS `directoryBrowser.ts:render()`: guide text + current-location label + path.
        let description = I18n.t("dir.guide") + "\n\n" + I18n.t("dir.current") + ": `\(current)`"

        let options: [WizardSelectOption]
        if children.isEmpty {
            options = [WizardSelectOption(label: I18n.t("dir.empty"), value: "__none__")]
        } else {
            options = children.map {
                WizardSelectOption(label: Self.clip($0, maxLabelLength), value: $0)
            }
        }
        let selectRow = WizardRow(components: [
            .select(
                customId: "dir:into",
                placeholder: children.isEmpty ? I18n.t("dir.empty") : I18n.t("dir.select"),
                options: options
            ),
        ])
        // Row1 ≤5 (Discord limit): up · here · resume · create · cancel. Row2: manual · optional panel.
        let actionButtons: [WizardComponent] = [
            .button(
                customId: "dir:up",
                label: I18n.t("dir.up"),
                style: .secondary,
                disabled: !canGoUp()
            ),
            .button(customId: "dir:here", label: I18n.t("dir.here"), style: .success, disabled: false),
            .button(customId: "dir:resume", label: I18n.t("dir.resume"), style: .primary, disabled: false),
            .button(customId: "dir:create", label: I18n.t("dir.create"), style: .secondary, disabled: false),
            .button(customId: "cancel", label: I18n.t("wizard.cancel"), style: .secondary, disabled: false),
        ]
        var pathButtons: [WizardComponent] = [
            .button(customId: "dir:manual", label: I18n.t("dir.manual"), style: .secondary, disabled: false),
        ]
        if nativePanel {
            pathButtons.append(
                .button(customId: "dir:panel", label: I18n.t("dir.panel"), style: .secondary, disabled: false)
            )
        }
        return WizardView(title: title, description: description, rows: [
            selectRow,
            WizardRow(components: actionButtons),
            WizardRow(components: pathButtons),
        ])
    }

    private func canGoUp() -> Bool {
        let parent = (current as NSString).deletingLastPathComponent
        return parent != current && Self.confine(parent, roots: roots)
    }

    // MARK: Path helpers

    private static func confine(_ target: String, roots: [String]?) -> Bool {
        let resolved = resolvePath(target)
        guard let roots else { return true }
        return roots.contains { isWithin(root: $0, child: resolved) }
    }

    private static func isDirectory(_ path: String) -> Bool {
        var isDir: ObjCBool = false
        guard FileManager.default.fileExists(atPath: path, isDirectory: &isDir) else { return false }
        return isDir.boolValue
    }

    /// Node `path.resolve` analogue — absolute + collapse `.`/`..` lexically.
    private static func resolvePath(_ path: String) -> String {
        if (path as NSString).isAbsolutePath {
            return URL(fileURLWithPath: path).standardizedFileURL.path
        }
        let joined = (FileManager.default.currentDirectoryPath as NSString).appendingPathComponent(path)
        return URL(fileURLWithPath: joined).standardizedFileURL.path
    }

    private static func resolvePath(_ base: String, _ child: String) -> String {
        if (child as NSString).isAbsolutePath {
            return resolvePath(child)
        }
        return URL(fileURLWithPath: base).appendingPathComponent(child).standardizedFileURL.path
    }

    private static func clip(_ s: String, _ max: Int) -> String {
        guard s.count > max else { return s }
        return String(s.prefix(max - 1)) + "…"
    }
}

/// Result of `createChild` (modal mkdir).
public enum DirCreateResult: Equatable, Sendable {
    case ok(path: String)
    case invalidName
    case escaped
    case failed(String)
}

/// True when `name` is a safe single path segment for 📁 create (TS `isSafeFolderName`).
public func isSafeFolderName(_ name: String) -> Bool {
    if name.isEmpty { return false }
    if name == "." || name == ".." { return false }
    if name.contains("/") || name.contains("\\") { return false }
    if (name as NSString).isAbsolutePath { return false }
    // basename must match exactly (no trailing separators / weird segments).
    if (name as NSString).lastPathComponent != name { return false }
    return true
}

/// custom_ids owned by the folder browser (wizard routing + DabMain modal/panel).
public func isDirectoryBrowserCustomId(_ customId: String) -> Bool {
    switch customId {
    case "dir:into", "dir:up", "dir:here",
         "dir:create", "dir:manual", "dir:panel",
         "dir:resume":
        return true
    default:
        return false
    }
}

// MARK: - G-P1-06 favorites → browseRoots (TS app.ts / slashCommands)

/// Map global `AppConfig.favorites` → `DirectoryBrowser` `allowedRoots`.
///
/// TS: `browseRoots: config.favorites` then
/// `allowedRoots` only when `browseRoots.length > 0` (empty → unbounded, Fix 1).
/// Empty/whitespace-only entries are dropped so a blank favorite cannot confine the browser.
public func browseRoots(fromFavorites favorites: [String]) -> [String] {
    favorites
        .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        .filter { !$0.isEmpty }
}
